File: examples/android/CMeSdkJava/app/src/main/java/me/connect/sdk/java/sample/db/ResponseConverter.java
package me.connect.sdk.java.sample.db;

import androidx.room.TypeConverter;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import me.connect.sdk.java.message.StructuredMessageHolder.Response;

public class ResponseConverter {
    @TypeConverter
    public static List<Response> fromString(String value) {
        try {
            List<Response> responses = new ArrayList<>();
            JSONArray json = new JSONArray(value);
            for (int i = 0; i < json.length(); i++) {
                JSONObject entry = json.getJSONObject(i);
                Response response = new Response(entry.getString("text"), entry.getString("nonce"));
                responses.add(response);
            }
            return responses;
        } catch (Exception e) {
            return null;
        }
    }

    @TypeConverter
    public static String fromResponseList(List<Response> responses) {
        try {
            JSONArray json = new JSONArray();
            if (responses != null) {
                for (Response r : responses) {
                    JSONObject entry = new JSONObject();
                    entry.put("text", r.getText());
                    entry.put("nonce", r.getNonce());
                    json.put(entry);
                }
                return json.toString();
            }
            return null;
        } catch (JSONException e) {
            return null;
        }
    }

}

File: examples/android/CMeSdkJava/app_kt/src/main/kotlin/me/connect/sdk/java/samplekt/db/ResponseConverter.kt
package me.connect.sdk.java.samplekt.db

import androidx.room.TypeConverter
import me.connect.sdk.java.message.StructuredMessageHolder.Response
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject


class ResponseConverter {
    @TypeConverter
    fun fromString(value: String?): List<Response> = try {
        val responses = mutableListOf<Response>()
        val json = JSONArray(value)
        for (i in 0 until json.length()) {
            val entry = json.getJSONObject(i)
            val response = Response(entry.getString("text"), entry.getString("nonce"))
            responses.add(response)
        }
        responses
    } catch (e: Exception) {
        emptyList()
    }

    @TypeConverter
    fun fromResponseList(responses: List<Response>?): String {
        try {
            val json = JSONArray()
            if (responses != null) {
                responses.forEach { r ->
                    val entry = JSONObject()
                    entry.put("text", r.text)
                    entry.put("nonce", r.nonce)
                    json.put(entry)
                }
                return json.toString()
            }
            return "[]"
        } catch (e: JSONException) {
            return "[]"
        }
    }
}
